cmd/crank/beta/validate/validate.go

```diff
@@ -17,102 +17,132 @@ limitations under the License.
+	runtimeschema "k8s.io/apimachinery/pkg/runtime/schema"
+	celconfig "k8s.io/apiserver/pkg/apis/cel"
 
 	"github.com/crossplane/crossplane-runtime/pkg/errors"
 
 	"github.com/crossplane/crossplane/internal/controller/apiextensions/composite"
 )
 
-func newValidators(crds []*extv1.CustomResourceDefinition) (map[schema.GroupVersionKind][]validation.SchemaValidator, error) {
-	validators := map[schema.GroupVersionKind][]validation.SchemaValidator{}
+func newValidatorsAndStructurals(crds []*extv1.CustomResourceDefinition) (map[runtimeschema.GroupVersionKind][]*validation.SchemaValidator, map[runtimeschema.GroupVersionKind]*schema.Structural, error) {
+	validators := map[runtimeschema.GroupVersionKind][]*validation.SchemaValidator{}
+	structurals := map[runtimeschema.GroupVersionKind]*schema.Structural{}
 
 	for i := range crds {
 		internal := &ext.CustomResourceDefinition{}
 		if err := extv1.Convert_v1_CustomResourceDefinition_To_apiextensions_CustomResourceDefinition(crds[i], internal, nil); err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 
-		// Top-level and per-version schemas are mutually exclusive. Therefore, we will use both if they are present.
+		// Top-level and per-version schemas are mutually exclusive.
 		for _, ver := range internal.Spec.Versions {
 			var sv validation.SchemaValidator
 			var err error
 
-			gvk := schema.GroupVersionKind{
+			gvk := runtimeschema.GroupVersionKind{
 				Group:   internal.Spec.Group,
 				Version: ver.Name,
 				Kind:    internal.Spec.Names.Kind,
 			}
 
-			// Version specific validation rules
-			if ver.Schema != nil && ver.Schema.OpenAPIV3Schema != nil {
-				sv, _, err = validation.NewSchemaValidator(ver.Schema.OpenAPIV3Schema)
-				if err != nil {
-					return nil, err
-				}
+			var s *ext.JSONSchemaProps
+			switch {
+			case internal.Spec.Validation != nil:
+				s = internal.Spec.Validation.OpenAPIV3Schema
+			case ver.Schema != nil && ver.Schema.OpenAPIV3Schema != nil:
+				s = ver.Schema.OpenAPIV3Schema
+			default:
+				// TODO log a warning here, it should never happen
+				continue
+			}
 
-				validators[gvk] = append(validators[gvk], sv)
+			sv, _, err = validation.NewSchemaValidator(s)
+			if err != nil {
+				return nil, nil, err
 			}
 
-			// Top level validation rules
-			if internal.Spec.Validation != nil {
-				sv, _, err = validation.NewSchemaValidator(internal.Spec.Validation.OpenAPIV3Schema)
-				if err != nil {
-					return nil, err
-				}
+			validators[gvk] = append(validators[gvk], &sv)
 
-				validators[gvk] = append(validators[gvk], sv)
+			structural, err := schema.NewStructural(s)
+			if err != nil {
+				return nil, nil, err
 			}
+
+			structurals[gvk] = structural
 		}
 	}
 
-	return validators, nil
+	return validators, structurals, nil
 }
 
 // SchemaValidation validates the resources against the given CRDs
 func SchemaValidation(resources []*unstructured.Unstructured, crds []*extv1.CustomResourceDefinition, skipSuccessLogs bool) error {
```

Comment:
Here we could get at least an io.Writer and pass a nop writer in the tests if we don't care about its content. Later on we can refactor it to return a list of errors and have the print logic separately

Reply:
done